app/src/main/java/org/breezyweather/sources/CommonConverter.kt

```diff
@@ -673,48 +678,190 @@ private fun completeHalfDayFromHourlyList(
 
     val newHalfDay = initialHalfDay ?: HalfDay()
 
+    val extremeTemperature = if (newHalfDay.temperature?.temperature == null
+        || newHalfDay.temperature.apparentTemperature == null
+        || newHalfDay.temperature.windChillTemperature == null
+        || newHalfDay.temperature.wetBulbTemperature == null) {
+        getHalfDayTemperatureFromHourlyList(newHalfDay.temperature, halfDayHourlyList, isDay)
+    } else newHalfDay.temperature
+
+    val totalPrecipitation = if (newHalfDay.precipitation?.total == null) {
+        getHalfDayPrecipitationFromHourlyList(halfDayHourlyList)
+    } else newHalfDay.precipitation
+
+    val maxPrecipitationProbability = if (newHalfDay.precipitationProbability?.total == null) {
+        getHalfDayPrecipitationProbabilityFromHourlyList(halfDayHourlyList)
+    } else newHalfDay.precipitationProbability
+
+    val maxWind = if (newHalfDay.wind?.speed == null) {
+        getHalfDayWindFromHourlyList(halfDayHourlyList)
+    } else newHalfDay.wind
+
+    val avgCloudCover = newHalfDay.cloudCover ?: getHalfDayCloudCoverFromHourlyList(halfDayHourlyList)
+
+    val avgVisibility = halfDayHourlyList.mapNotNull { it.visibility }.let {
+        if (it.isNotEmpty()) it.average() else null
+    }
+
     var halfDayWeatherText = newHalfDay.weatherText
     var halfDayWeatherPhase = newHalfDay.weatherPhase
     var halfDayWeatherCode = newHalfDay.weatherCode
 
     // Weather code + Weather text
     if (halfDayWeatherCode == null || halfDayWeatherText == null) {
```

Comment:
Can you move icon computing to a separate function? Thanks!

Reply:
Done.